### PR TITLE
adds support for FQCN form type referencing to FormContractor

### PR DESCRIPTION
**Pull request by ju1ius**
I am targetting this branch, because this is a BC bugfix

Will close https://github.com/sonata-project/SonataAdminBundle/issues/3976 when [doctrine-mongodb-admin-bundle PR](https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/pull/178) is also merged.

## Changelog

```markdown
### Fixed
- Fix `FormContractor::getDefaultOptions` not checking against form types FQCNs
```

## Subject

`FormContractor::getDefaultOptions` didn't check against form types FQCNs, causing the default options not to be set when using the Symfony 3 way:
```php
$formMapper->add('foo', ModelType::class);
```
See related admin-bundle issue at https://github.com/sonata-project/SonataAdminBundle/issues/3976

### CODE REVIEW

**Comment by greg0ire** [comment]
LVGTM (looks very good to merge :+1: )

File: Tests/Builder/FormContractorTest.php [hunk 1]
```diff
@@ -46,4 +47,68 @@ public function testGetFormBuilder()
             $this->formContractor->getFormBuilder('test', array('foo' => 'bar'))
         );
     }
+
+    public function testDefaultOptionsForSonataFormTypes()
+    {
+        $admin = $this->getMockBuilder('Sonata\AdminBundle\Admin\AdminInterface')->getMock();
```
Comment (greg0ire) [inline Tests/Builder/FormContractorTest.php]: I think you can do this directly : `$this->getMock('Sonata\AdminBundle\Admin\AdminInterface');`

Comment (ju1ius) [inline Tests/Builder/FormContractorTest.php]: Yep, but it's deprecated.

Comment (greg0ire) [inline Tests/Builder/FormContractorTest.php]: Ooooh ok I thought the only other solution was `createMock`, that's great! What do you think @Soullivaneuh ? Should we go that way or not?

Comment (ju1ius) [inline Tests/Builder/FormContractorTest.php]: This way produces less warning noise in the build output.

Comment (greg0ire) [inline Tests/Builder/FormContractorTest.php]: It's indeed worrying that other problems might not be seen because of this.

Comment (ju1ius) [inline Tests/Builder/FormContractorTest.php]: That's the only way I found to remove deprecation warnings while supporting both phpunit 4 and 5.
Don't know if there are others...

Comment (greg0ire) [inline Tests/Builder/FormContractorTest.php]: I don't remember why we don't use phpunit 4.8 for every build. Help @sonata-project/contributors !

Comment (ju1ius) [inline Tests/Builder/FormContractorTest.php]: PHP 7.0 ?

Comment (greg0ire) [inline Tests/Builder/FormContractorTest.php]: Not sure why (can't see a conflict in [the Composer manifest](https://packagist.org/packages/phpunit/phpunit#4.8.26) ) but that does feel oddly familiar…

Comment (Soullivaneuh) [inline Tests/Builder/FormContractorTest.php]: I would prefer `getMock` because the mock builder for an interface is useless.

> This way produces less warning noise in the test/build output.

Yes, and the mock builder will be still like this even after PHPUnit 4 support dropping because of that.

It produce deprecated warning but for now we have no choice.

**Comment by Soullivaneuh** [comment]
Looks great! I'll do a manual check ASAP.

File: Tests/Builder/FormContractorTest.php (hunk 1 above)
Comment (Soullivaneuh) [inline Tests/Builder/FormContractorTest.php]: > I don't remember why we don't use phpunit 4.8 for every build.

PHPUnit 4.8 is not **fully** compatible with PHP7. You can have some not fixed edge case I think.

Comment (greg0ire) [inline Tests/Builder/FormContractorTest.php]: Ok so that's why… too bad.

File: Builder/FormContractor.php [hunk 2]
```diff
@@ -98,20 +98,32 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
...
+        ), true)) {
+            if ($fieldDescription->getOption('edit') === 'list') {
                 throw new \LogicException('The ``sonata_type_model`` type does not accept an ``edit`` option anymore, please review the UPGRADE-2.1.md file from the SonataAdminBundle');
             }
 
             $options['class'] = $fieldDescription->getTargetEntity();
             $options['model_manager'] = $fieldDescription->getAdmin()->getModelManager();
 
-            if ($type == 'sonata_type_model_autocomplete') {
+            if ($type === 'sonata_type_model_autocomplete'
```
Comment (Soullivaneuh) [inline Builder/FormContractor.php]: You can use `in_array` for that.

File: Builder/FormContractor.php [hunk 3]
```diff
@@ -98,20 +98,32 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
...
+                || $type === 'Sonata\AdminBundle\Form\Type\ModelAutocompleteType'
+            ) {
                 if (!$fieldDescription->getAssociationAdmin()) {
                     throw new \RuntimeException(sprintf('The current field `%s` is not linked to an admin. Please create one for the target entity: `%s`', $fieldDescription->getName(), $fieldDescription->getTargetEntity()));
                 }
             }
-        } elseif ($type == 'sonata_type_admin') {
+        } elseif ($type === 'sonata_type_admin' || $type === 'Sonata\AdminBundle\Form\Type\AdminType') {
```
Comment (Soullivaneuh) [inline Builder/FormContractor.php]: You can use `in_array` for that.

File: Builder/FormContractor.php [hunk 4]
```diff
@@ -126,7 +138,7 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
 
             $options['data_class'] = $fieldDescription->getAssociationAdmin()->getClass();
             $fieldDescription->setOption('edit', $fieldDescription->getOption('edit', 'admin'));
-        } elseif ($type == 'sonata_type_collection') {
+        } elseif ($type === 'sonata_type_collection' || $type === 'Sonata\CoreBundle\Form\Type\CollectionType') {
```
Comment (Soullivaneuh) [inline Builder/FormContractor.php]: You can use `in_array` for that.

File: Tests/Builder/FormContractorTest.php [hunk 5]
```diff
@@ -31,7 +32,7 @@
 
     protected function setUp()
     {
-        $this->formFactory = $this->getMock('Symfony\Component\Form\FormFactoryInterface');
+        $this->formFactory = $this->getMockBuilder('Symfony\Component\Form\FormFactoryInterface')->getMock();
```
Comment (Soullivaneuh) [inline Tests/Builder/FormContractorTest.php]: Use `getMock` for interfaces.

File: Tests/Builder/FormContractorTest.php [hunk 6]
```diff
@@ -46,4 +47,69 @@ public function testGetFormBuilder()
             $this->formContractor->getFormBuilder('test', array('foo' => 'bar'))
         );
     }
+
+    public function testDefaultOptionsForSonataFormTypes()
+    {
+        $admin = $this->getMockBuilder('Sonata\AdminBundle\Admin\AdminInterface')->getMock();
+        $modelManager = $this->getMockBuilder('Sonata\AdminBundle\Model\ModelManagerInterface')->getMock();
```
Comment (Soullivaneuh) [inline Tests/Builder/FormContractorTest.php]: Use `getMock` for interfaces.

File: Tests/Builder/FormContractorTest.php [hunk 7]
```diff
@@ -46,4 +47,69 @@ public function testGetFormBuilder()
...
+        $modelClass = 'FooEntity';
+
+        $admin->method('getModelManager')->willReturn($modelManager);
+        $admin->method('getClass')->willReturn($modelClass);
+
+        $fieldDescription = $this->getMockBuilder('Sonata\AdminBundle\Admin\FieldDescriptionInterface')->getMock();
```
Comment (Soullivaneuh) [inline Tests/Builder/FormContractorTest.php]: Use `getMock` for interfaces.

**Comment by Soullivaneuh** [comment]
Did a manual check, works great! :+1: 

Please update the code according to our comments and the PR will be ready to merge.

**Comment by Soullivaneuh** [comment]
Please see #598
